pkg/util/keymutex/keymutex.go

```diff
@@ -58,26 +66,36 @@ func (km *keyMutex) LockKey(id string) {
 func (km *keyMutex) UnlockKey(id string) error {
 	glog.V(5).Infof("UnlockKey(...) called for id %q\r\n", id)
 	km.RLock()
-	defer km.RUnlock()
 	mutex, exists := km.mutexMap[id]
 	if !exists {
+		km.RUnlock()
 		return fmt.Errorf("id %q not found", id)
 	}
 	glog.V(5).Infof("UnlockKey(...) for id. Mutex found, trying to unlock it. %q\r\n", id)
 
 	mutex.Unlock()
 	glog.V(5).Infof("UnlockKey(...) for id %q completed.\r\n", id)
+	km.RUnlock()
+
+	km.Lock()
```

Comment:
This is a race - a parallel call to UnlockKey could win and remove the item.  This would then be a nil dereference.

I took a quick look at users of this, and honestly I am not convinced that this abstraction is necessary at all.  I bet that we could just make it go away, use a simpler lock in the plugins that need it, and nobody would notice.

But mostly the plugins don't document WHY they need this, so it's really hard to tell if this would matter.

Reply:
A parallel call to UnlockKey could win but results will be same IMO, because number of goroutines to call UnlockKey to decrease refcnt cannot be more than gorountines hold the lock, order of goroutines to deference refcnt does not matter here. If users call an extra call to UnlockKey if code wrong, program will panic first at [`mutex.Lock`](https://golang.org/pkg/sync/#Mutex.Unlock) because it is a run-time error to unlock a already-unlocked lock.

I did some searching, and found  `keymutex` library was first introduced in #16053, and in [this tracking issue](https://github.com/kubernetes/kubernetes/issues/16305#issuecomment-229476219), @saad-ali said:

> With `goroutinemap` and `operationexecutor`, `keymutex` is no longer needed. All of this code can be removed.

@msau42 Most plugins does not use `keymutex` at all (e.g. rbd), is `keymutex` still be needed in `local` plugin?

Reply:
The local plugin uses it to serialize the fsGroup conflict check, if multiple pods share the same volume.

Because the local plugin doesn't implement the Attacher interface, MountVolume can be called in [parallel](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L744).  However, there is work being done to introduce a [Local Attacher](https://github.com/kubernetes/community/pull/1683) volume interface, and move local volumes to implement that.  When that happens, then MountVolume will be serialized per volume, and then we should not need this lock anymore.

I'm not sure about the other plugins, but I think they don't implement Attacher either, and use the lock to serialize mount refcounting.